src/optimizer/optimizer_context.py

```diff
@@ -0,0 +1,93 @@
+# coding=utf-8
+# Copyright 2018-2020 EVA
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import copy
+from src.optimizer.optimizer_task_stack import OptimizerTaskStack
+from src.optimizer.memo import Memo
+from src.optimizer.operators import Operator
+from src.optimizer.group_expression import GroupExpression
+from src.optimizer.group import INVALID_GROUP_ID
+
+
+class OptimizerContext:
+    """
+        Maintain context information for the optimizer
+
+        Arguments:
+            _task_queue(OptimizerTaskStack):
+                stack to keep track outstanding tasks
+    """
+
+    def __init__(self):
+        self._task_stack = OptimizerTaskStack()
+        self._memo = Memo()
+
+    @property
+    def task_stack(self):
+        return self._task_stack
+
+    @property
+    def memo(self):
+        return self._memo
+
+    def xform_opr_to_group_expr(
+        self,
+        opr: Operator,
+        root_group_id: int = INVALID_GROUP_ID,
+        is_root: bool = False,
+        copy_opr: bool = True
+    ) -> GroupExpression:
+        """
+        Generate a group expression from a logical operator.
```

Comment:
Also related to the last discussion.

There are two cases that root id cannot be guaranteed. 
1. Group ID is just a list index, so I think `group_id <= len(group)`. Otherwise, there will be an empty group in the middle.
2. Again if memo is empty, the Group ID cannot be guaranteed.

Reply:
The time you want to guarantee a root id is when a rewrite happens. So there must be group expressions in the memo. Neither of the cases you mentioned should happen if optimizer works as intended.